misc/python/materialize/cli/mzcompose.py

```diff
@@ -177,142 +189,251 @@ def gen_shortcuts(repo: mzbuild.Repository) -> int:
+
+    def configure(self, parser: argparse.ArgumentParser) -> None:
+        parser.add_argument("service", metavar="SERVICE", help="the service to target")
+
+    def run(self, args: argparse.Namespace) -> None:
+        super().run(args)
+        composition = load_composition(args)
+        ports = composition.find_host_ports(args.service)
+        if len(ports) == 1:
+            webbrowser.open(f"http://localhost:{ports[0]}")
+        elif not ports:
+            raise errors.MzRuntimeError(f"No running services matched {args.service!r}")
+        else:
+            raise errors.MzRuntimeError(
+                f"Too many ports matched {args.service!r}, found: {ports}"
+            )
 
 
-# We subclass `argparse.ArgumentParser` so that we can override its default
-# behavior of exiting on error. We want Docker Compose to be responsible for
-# generating option-parsing errors.
-class ArgumentParser(argparse.ArgumentParser):
-    def __init__(self) -> None:
-        super().__init__(add_help=False)
-        # If you add an option here, add it to the help text in `print_help`.
-        self.add_argument("--mz-quiet", action="store_true", default=None)
-        self.add_argument("--mz-find")
-        self.add_argument(
-            "--mz-build-mode", default="release", choices=["dev", "release"]
+class DockerComposeCommand(Command):
+    add_help = False
+
+    def __init__(self, name: str, help: str, acquire_deps: bool = False):
+        self.name = name
+        self.help = help
+        self.acquire_deps = acquire_deps
+
+    def configure(self, parser: argparse.ArgumentParser) -> None:
+        parser.add_argument("-h", "--help", action="store_true")
+
+    def run(self, args: argparse.Namespace) -> None:
+        if args.help:
+            output = spawn.capture(
+                ["docker-compose", self.name, "--help"], stderr_too=True, unicode=True
+            )
+            output = output.replace("docker-compose", "./mzcompose")
+            print(output, file=sys.stderr)
+            return
+
+        # Make sure Docker Compose is new enough.
+        output = (
+            spawn.capture(["docker-compose", "version", "--short"], unicode=True)
+            .strip()
+            .strip("v")
+        )
+        version = tuple(int(i) for i in output.split("."))
+        if version < MIN_COMPOSE_VERSION:
+            msg = f"Unsupported docker-compose version: {version}, min required: {MIN_COMPOSE_VERSION}"
+            raise errors.MzConfigurationError(msg)
+
+        announce("Collecting mzbuild dependencies")
+        composition = load_composition(args)
+        deps = composition.repo.resolve_dependencies(composition.images)
+        for d in deps:
+            say(d.spec())
+
+        if self.acquire_deps:
+            if args.coverage:
+                # If the user has requested coverage information, create the
+                # coverage directory as the current user, so Docker doesn't create
+                # it as root.
```

Comment:
random q, do we subprocess as root or is the docker daemon handling this stuff?

Reply:
Yeah it's the Docker daemon. The client runs as you, but connects to the Docker daemon which does arbitrary stuff (like making volumes) as root.